src/server/accounting.c

```diff
@@ -353,244 +390,138 @@ get_walltime(job *jp, int res)
  * @par	Functionality:
  *	Used by account_jobstr() and account_jobend()
  *
- *
  * @param[in]	pjob	- pointer to job structure
- * @param[in]	type	- account record type
  * @param[in]	buf	- buffer holding the data that will be stored in
  *			  accounting logs.
- * @param[in]	len	- number of characters in 'buf' still available to
- *			  store data.
- * @return	char *
- * @retval	pointer to 'buf' containing new data.
+ * @param[in]	len	- pointer to the size of 'buf'.
+ * @return	void
  *
  */
-static char *
-acct_job(job *pjob, int type, char *buf, int len)
+static void
+acct_job(job *pjob, int type, char **buf, int *len)
 {
-	pbs_list_head attrlist;
-	int	  i, k;
-	int       nd;
-	svrattrl *pal;
-	char *pb;
-	int	att_index;
-	int	len_orig;
-	char	save_char;
-	int	old_perm;
+	pbs_list_head 	attrlist;
+	svrattrl 	*pal = NULL;
+#ifdef WIN32
+	char		*win_str = NULL;
+#endif
+	char 		timestr[ACCTBUF_TIMES_NEED] = {0};
+	char            save_char = 0;
+	int		att_index;
+	int		k = 0;
+	int		old_perm;
+	int		len_orig;
 
-	pb = buf;
 	CLEAR_HEAD(attrlist);
 
 	/* gridname */
 	if (pjob->ji_wattr[(int)JOB_ATR_gridname].at_flags & ATR_VFLAG_SET) {
-		nd = strlen(pjob->ji_wattr[(int)JOB_ATR_gridname].at_val.at_str)
-			+ sizeof(GRIDNAME_FMT);
-		if (nd > len)
-			if (grow_acct_buf(&pb, &len, nd) == -1)
-				return (pb);
-
-		(void)snprintf(pb, len, GRIDNAME_FMT,
-			pjob->ji_wattr[(int)JOB_ATR_gridname].at_val.at_str);
-		i = strlen(pb);
-		pb  += i;
-		len -= i;
+		if (!pbs_strcat(buf, len, GRIDNAME_FMT)) 
```

Comment:
I don't like that we have modified format string because we now rely on pbs_strcat to grow the buffer instead of grow_acct_buf. I think is not efficient to call pbs_strcat for small strings (like space char)
It is probably better to write a new function which is a mix of asprintf() and strcat(). This function should be able to grow the buffer for you like pbs_strcat() does and should also be able to take a formatted string input as well.
what do you think?

Reply:
asprintf() is a GNU-extension so it can create issue of platforms and may not work on windows. Because of which I have used snprintf() to many places where it can be used and pbs_strcat() as well.

Reply:
Use pbs_asprintf() as defined in misc_utils.c. That will get you around the portability issue.

Reply:
The idea which you are saying it is possible but that will create lot of complexity as asprintf() will always overwrite the buffer and here in code we are adding in the buffer after every few lines so to make it possible we need to pass buffer+place from where it should start so that it will not write in the buffer from the start again. So for that this calculation should be made on every point that from were it should start adding the data in the buffer as I mentioned we are adding into buffer after every few lines.